### PR TITLE
Query JPA repository with single table inheritance subtypes

### PR DESCRIPTION
**Pull request by mee-omi**

### CODE REVIEW

File: persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/entity/Employee.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    public Employee(String name, int age) {
+        super();
+        this.name = name;
+        this.age = age;
+    }
+
+    public Employee() {
+        super();
+        // TODO Auto-generated constructor stub
```
Comment (theangrydev) [inline persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/entity/Employee.java:56]:
```suggestion
```

File: persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/config/PersistenceConfig.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+@PropertySource("classpath:persistence.properties")
+@EnableTransactionManagement
+public class PersistenceConfig {
+
+    @Autowired
+    private Environment env;
+
+    @Bean
+    public LocalContainerEntityManagerFactoryBean entityManagerFactory() {
+        final LocalContainerEntityManagerFactoryBean em = new LocalContainerEntityManagerFactoryBean();
```
Comment (theangrydev) [inline persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/config/PersistenceConfig.java:33]: Remove unnecessary final usage (same elsewhere)
```suggestion
        LocalContainerEntityManagerFactoryBean em = new LocalContainerEntityManagerFactoryBean();
```

File: persistence-modules/spring-jpa-3/src/test/java/com/baeldung/jpa/subtypes/JPAInheritanceSingleTableTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import com.baeldung.jpa.subtypes.entity.ContractEmployee;
+import com.baeldung.jpa.subtypes.entity.Employee;
+import com.baeldung.jpa.subtypes.entity.PermanentEmployee;
+import com.baeldung.jpa.subtypes.repository.ContractEmployeeRepository;
+import com.baeldung.jpa.subtypes.repository.EmployeeRepository;
+import com.baeldung.jpa.subtypes.repository.PermanentEmployeeRepository;
+
+@ExtendWith(SpringExtension.class)
+@SpringBootTest(classes = SubTypeJPAApplication.class)
+class JPAInheritanceSingleTableTest {
```
Comment (theangrydev) [inline persistence-modules/spring-jpa-3/src/test/java/com/baeldung/jpa/subtypes/JPAInheritanceSingleTableTest.java:24]: We follow https://google.github.io/styleguide/javaguide.html for naming conventions
```suggestion
class JpaInheritanceSingleTableTest {
```

File: persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/SubTypeJPAApplication.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.domain.EntityScan;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
+
+@SpringBootApplication 
+@EnableJpaRepositories("com.baeldung.jpa.subtypes.repository")
+public class SubTypeJPAApplication {
```
Comment (theangrydev) [inline persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/SubTypeJPAApplication.java:11]: We follow https://google.github.io/styleguide/javaguide.html for naming conventions
```suggestion
public class SubTypeJpaApplication {
```

File: persistence-modules/spring-jpa-3/src/test/java/com/baeldung/jpa/subtypes/JPAInheritanceSingleTableTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+    
+    @Test
+    void givenRecordsInserted_whenAllEmployeesQueried_thenAllEmployeesRecordsRetruned() {
+        List<Employee> empList = empRepository.findAll();
+        assertThat(empList.size()).isEqualTo(6);
+    }
+    
+    @Test
+    void givenRecordsInserted_whenDiscrimnatorQueryUsed_thenEmployeesRecordsByTypeRetruned() {
```
Comment (theangrydev) [inline persistence-modules/spring-jpa-3/src/test/java/com/baeldung/jpa/subtypes/JPAInheritanceSingleTableTest.java:80]: Fix typo (same in other tests)
```suggestion
    void givenRecordsInserted_whenDiscrimnatorQueryUsed_thenEmployeesRecordsByTypeReturned() {
```

File: persistence-modules/spring-jpa-3/src/test/java/com/baeldung/jpa/subtypes/JPAInheritanceSingleTableTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        ContractEmployee contr3 = new ContractEmployee(365, "Greg", 41);
+
+        contrEmpRepository.save(contr1);
+        contrEmpRepository.save(contr2);
+        contrEmpRepository.save(contr3);
+
+    }
+
+    @AfterEach
+    public void tearDown() {
```
Comment (theangrydev) [inline persistence-modules/spring-jpa-3/src/test/java/com/baeldung/jpa/subtypes/JPAInheritanceSingleTableTest.java:56]:
```suggestion
    void tearDown() {
```

**Review by theangrydev — CHANGES_REQUESTED** [review]

File: persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/config/PersistenceConfig.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+        return transactionManager;
+    }
+
+    @Bean
+    public PersistenceExceptionTranslationPostProcessor exceptionTranslation() {
+        return new PersistenceExceptionTranslationPostProcessor();
+    }
+
+    final Properties additionalProperties() {
```
Comment (theangrydev) [inline persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/config/PersistenceConfig.java:68]: Remove unnecessary final + make private if possible 
```suggestion
    private Properties additionalProperties() {
```

File: persistence-modules/spring-jpa-3/src/test/java/com/baeldung/jpa/subtypes/SubTypeJpaImplTest.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+import com.baeldung.jpa.subtypes.entity.ContractEmployee;
+import com.baeldung.jpa.subtypes.entity.Employee;
+import com.baeldung.jpa.subtypes.entity.PermanentEmployee;
+import com.baeldung.jpa.subtypes.repository.ContractEmployeeRepository;
+import com.baeldung.jpa.subtypes.repository.EmployeeRepository;
+import com.baeldung.jpa.subtypes.repository.PermanentEmployeeRepository;
+
+@ExtendWith(SpringExtension.class)
+@SpringBootTest(classes = SubTypeJpaApplication.class)
+class SubTypeJpaImplTest {
```
Comment (theangrydev) [inline persistence-modules/spring-jpa-3/src/test/java/com/baeldung/jpa/subtypes/SubTypeJpaImplTest.java:24]: I prefer the old name, it's clearer
```suggestion
class JpaInheritanceSingleTableTest {
```

**Review by theangrydev — CHANGES_REQUESTED** [review]

**Review by theangrydev — APPROVED** [review]

File: persistence-modules/spring-jpa-3/pom.xml [hunk 9]
```diff
@@ -1,48 +1,52 @@
...
-    <version>0.1-SNAPSHOT</version>
-    <name>spring-jpa-3</name>
-    <packaging>war</packaging>
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<artifactId>spring-jpa-3</artifactId>
+	<version>0.1-SNAPSHOT</version>
+	<name>spring-jpa-3</name>
+	<packaging>war</packaging>
```
Comment (KevinGilmore) [inline persistence-modules/spring-jpa-3/pom.xml:9]: We need to revert the spacing to use spaces, not tabs. There should not be this many changes to the pom.xml.

File: persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/repository/EmployeeRepository.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.data.jpa.repository.Query;
+
+import com.baeldung.jpa.subtypes.entity.ContractEmployee;
+import com.baeldung.jpa.subtypes.entity.Employee;
+import com.baeldung.jpa.subtypes.entity.PermanentEmployee;
+
+public interface EmployeeRepository extends JpaRepository<Employee, Long> {
+
+    @Query("SELECT e FROM Employee e WHERE type(e) = 'PERM'")
+    List<PermanentEmployee> findAllPermEmplouees();
```
Comment (KevinGilmore) [inline persistence-modules/spring-jpa-3/src/main/java/com/baeldung/jpa/subtypes/repository/EmployeeRepository.java:15]:
```suggestion
    List<PermanentEmployee> findAllPermEmployees();
```

**Pull request merged**